src/net/ossl.cc

```diff
@@ -1178,15 +1202,17 @@ class session : public enable_shared_from_this<session>, public session_impl {
         return ssl_ctx;
     }
 
-    static std::optional<sstring> get_ossl_string(X509_NAME* name){
-        if (auto name_str = X509_NAME_oneline(name, nullptr, 0)) {
-            // sstring constructor may throw, to ensure deallocation of this OpenSSL string in
-            // all cases, wrap the call to free() in a deferred_action
-            auto done = defer([&name_str]() noexcept { OPENSSL_free(name_str); });
-            sstring ossl_str(name_str);
-            return ossl_str;
+    static std::optional<sstring> get_dn_string(X509_NAME* name) {
+        auto out = bio_ptr(BIO_new(BIO_s_mem()));
+        if (!X509_NAME_print_ex(out.get(), name, 0, ASN1_STRFLGS_RFC2253 | XN_FLAG_SEP_COMMA_PLUS |
+                                XN_FLAG_FN_SN | XN_FLAG_DUMP_UNKNOWN_FIELDS)) {
+            return std::nullopt;
         }
-        return std::nullopt;
+        static const auto max_len = 4096;
+        char buffer[max_len];
+        memset(buffer, 0, max_len);
+        BIO_read(out.get(), buffer, max_len - 1);
+        return sstring(buffer);
```

Comment:
```suggestion
        char * bio_ptr;
        auto len = BIO_get_mem_data(out.get(), &bio_ptr);
        return sstring(bio_ptr, len);
```

Reply:
>        char * bio_ptr = nullptr;
>        auto len = BIO_get_mem_data(out.get(), &bio_ptr);
>        return sstring(bio_ptr, len);

> BIO_get_mem_data() returns the total number of bytes available on success, 0 if b is NULL, or a negative value in case of other errors.

Looks like the error condition should be checked?